### ISSUE TITLE
Adding support for vSphereVolume

### ISSUE THREAD

**Issue by snsumner** [upmc-enterprises/elasticsearch-operator#100]
It would be quite useful if the Elasticsearch operator add support for using vSphereVolume or NFS. I've tried tweaking it to work but have no success. 

```
[Wed Sep 20 23:32:47 core@jumpbox:~/github/snsumner/elasticsearch-operator-helm2 ] $ kubectl describe pvc es-data-es-data-standard-0 -n elasticsearch
Name:		es-data-es-data-standard-0
Namespace:	elasticsearch
StorageClass:	standard
Status:		Pending
Volume:
Labels:		component=elasticsearch
		name=es-data-standard
		role=data
Annotations:	volume.beta.kubernetes.io/storage-class=standard
		volume.beta.kubernetes.io/storage-provisioner=kubernetes.io/vsphere-volume
Capacity:
Access Modes:
Events:
  FirstSeen	LastSeen	Count	From				SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  2m		56s		2	persistentvolume-controller			Warning		ProvisioningFailed	Failed to provision volume with StorageClass "standard": invalid option "zone" for volume plugin kubernetes.io/vsphere-volume
  2m		11s		10	persistentvolume-controller			Warning		ProvisioningFailed	Failed to provision volume with StorageClass "standard": invalid option "type" for volume plugin kubernetes.io/vsphere-volume```

**Comment by stevesloka** [comment]
What does your TPR look like? Would you be able to share that? Could you share the storage class that got created?

**Comment by snsumner** [comment]
This is what my hacked TPR looks like:

````
[Thu Sep 21 10:57:51 core@jumpbox:~/github/snsumner/elasticsearch-operator-helm2/templates ] $ cat elasticsearchcluster.yaml
{{- if not .Release.IsInstall -}}
apiVersion: enterprises.upmc.com/v1
kind: ElasticsearchCluster
metadata:
  name: {{ .Values.es_cluster_name }}
  namespace: {{ .Values.namespace }}
  labels:
    chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
spec:
  master-node-replicas: {{ .Values.es_master_replicas }}
  client-node-replicas: {{ .Values.es_client_replicas }}
  data-node-replicas: {{ .Values.es_data_replicas }}
  data-volume-size: {{ .Values.volume_size }}
  elastic-search-image: "{{ .Values.es_image }}:{{ .Values.es_tag }}"
  java-options: "-Xms{{ .Values.java_xms }} -Xmx{{ .Values.java_xmx }}"
  smapshot:
    scheduler-enabled: {{ .Values.schedule_snapshot }}
    bucket-name: {{ .Values.bucket_name }}
    cron-schedule: '{{ .Values.cron_schedule }}'
  storage:
    storage-class-provisioner: kubernetes.io/vsphere-volume
    type: fast
  zones: []
{{- end -}}
```
I've tried with or without Type defined.

This is what the storage class that got created looks like:

```
[Thu Sep 21 10:56:19 core@jumpbox:~/github/snsumner/elasticsearch-operator-helm2 ] $ kubectl get sc
NAME                 TYPE
etcd-backup-gce-pd   kubernetes.io/gce-pd
fast (default)       kubernetes.io/vsphere-volume
standard             kubernetes.io/vsphere-volume
[Thu Sep 21 10:56:23 core@jumpbox:~/github/snsumner/elasticsearch-operator-helm2 ] $ kubectl describe sc standard
Name:		standard
IsDefaultClass:	No
Annotations:	<none>
Provisioner:	kubernetes.io/vsphere-volume
Parameters:	type=fast,zone=standard
Events:		<none>
```

**Comment by stevesloka** [comment]
I think the `type` and `zone` piece is part of the problem. Let me see if I can work that out into params or something. 

I don't have access to vSphere, but I do have ESXi. Can I test on that? 

Also, you mentioned the same with NFS? I do have an NFS server I could spin up to validate on.

**Comment by snsumner** [comment]
I came to similar conclusion that `type` and `zone` are not valid parameters for the vSphereVolume storage class.  

Using ESXi alone will not allow you to use vSphereVolume. The reason is that the setup requires the storage class to be able to access the vSphere server's API.

If your interested I can give you temporary access to my homelab sandbox environment where I have vSphereVolume setup to debug. You can ping me at scott@coreos.com if you want to take me up on this offer.

The reason I would like your operator to support vSphereVolume is that we run into a number of customers that want to use or evaluate Tectonic on-premise and typically they do it on VMware because getting bare metal is more time consuming. It would be nice to have an offering that supported a persistent volume that is commonly used on-premise. The ones I typically see are vSphereVolumes, NFS and iSCSI.

I also have NFS setup in my homelab but I've been told that I should avoid using NFS if at all possible with Kubernetes. Another option I've been suggested to look at is GlusterFS but this is more effort to setup then using vSphereVolume.

**Comment by stevesloka** [comment]
Hey @snsumner I haven't forgotten about this, going to carve out some time today to get this working for you.

**Comment by stevesloka** [comment]
TLDR: Scott was able to create a storage class using the VMWare volumes, we referenced that from the operator and the volumes were provisioned dynamically.
